pkg/webhook/resources/upgrade/validator.go

```diff
@@ -307,6 +312,44 @@ func (v *upgradeValidator) checkMachines() error {
 	return nil
 }
 
+func (v *upgradeValidator) checkSingleReplicaVolumes() error {
+	// Upgrade should be rejected if any single-replica volume exists
+	nodes, err := v.nodes.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	// Skip for single-node cluster
+	if len(nodes) == 1 {
+		return nil
+	}
+
+	// Find all single-replica volumes
+	singleReplicaVolumes, err := v.lhVolumes.GetByIndex(indexeres.VolumeByReplicaCountIndex, "1")
+	if err != nil {
+		return err
+	}
+
+	volumeNames := make([]string, 0, len(singleReplicaVolumes))
+	for _, volume := range singleReplicaVolumes {
+		// Ignore detached volumes
+		if volume.Status.State == "detached" {
```

Comment:
nit: should we also skip the `deleting`?

Reply:
Good call! I guess we only need to take care of creating, attached, and attaching?

```go
const (
	VolumeStateCreating  = VolumeState("creating")
	VolumeStateAttached  = VolumeState("attached")
	VolumeStateDetached  = VolumeState("detached")
	VolumeStateAttaching = VolumeState("attaching")
	VolumeStateDetaching = VolumeState("detaching")
	VolumeStateDeleting  = VolumeState("deleting")
)
```

Reply:
Yes, I think so. `detaching`, `detached`, and `deleting` could be ignored.

Reply:
Updated and PTAL, thanks!

Reply:
Looks good! Thanks!!